### ISSUE TITLE
Third-party npm xmllint causes Meteor restart

### ISSUE THREAD

**Issue by CTimmerman** [meteor/meteor#7012]
**Repo:** https://github.com/CTimmerman/meteor-xmllint-bug/

[This simple testcase ](https://github.com/CTimmerman/meteor-xmllint-bug/blob/master/test.js) fails:

```
C:\code\github>git clone git@github.com:CTimmerman/meteor-xmllint-bug.git
Cloning into 'meteor-xmllint-bug'...
Enter passphrase for key '/c/Users/Cees.Timmerman/.ssh/id_rsa':
remote: Counting objects: 33, done.
remote: Total 33 (delta 0), reused 0 (delta 0), pack-reused 32
Receiving objects: 100% (33/33), 775.87 KiB | 663.00 KiB/s, done.
Resolving deltas: 100% (3/3), done.
Checking connectivity... done.

C:\code\github>cd meteor-xmllint-bug

C:\code\github\meteor-xmllint-bug>meteor
[[[[[ C:\code\github\meteor-xmllint-bug ]]]]]

=> Started proxy.
=> Meteor 1.3.2.4 is available. Update this project with 'meteor update'.
=> Started MongoDB.
npm-container: updating npm dependencies -- xmllint...
=> Started your app.

=> App running at: http://localhost:3000/
   Type Control-C twice to stop.

I20160509-18:47:50.146(2)?
I20160509-18:47:50.147(2)? RESULT { errors: [ 'file_0.xml:2: element Article: Schemas validity error : Element \'Article\': No matching global declaration available for the validation root.' ] }
=> Exited with code: 3
Terminate batch job (Y/N)?
^C

```

```
C:\code\github\meteor-xmllint-bug>meteor update

...
snip
...
* These packages have been updated to new versions that are not backwards
compatible.
meteor-xmllint-bug: updated to Meteor 1.3.2.4.

Changes to your project's package version selections from updating package versions:

standard-minifiers  removed from your project
```

```
C:\code\github\meteor-xmllint-bug>meteor
[[[[[ C:\code\github\meteor-xmllint-bug ]]]]]

=> Started proxy.
=> Started MongoDB.
=> Started your app.

=> App running at: http://localhost:3000/
   Type Control-C twice to stop.

I20160509-18:49:23.843(2)?
I20160509-18:49:23.844(2)? RESULT { errors: [ 'file_0.xml:2: element Article: Schemas validity error : Element \'Article\': No matching global declaration available for the validation root.' ] }
=> Exited with code: 3
I20160509-18:49:27.721(2)?
I20160509-18:49:27.728(2)? RESULT { errors: [ 'file_0.xml:2: element Article: Schemas validity error : Element \'Article\': No matching global declaration available for the validation root.' ] }
=> Meteor server restarted
=> Exited with code: 3
I20160509-18:49:32.118(2)?
I20160509-18:49:32.125(2)? RESULT { errors: [ 'file_0.xml:2: element Article: Schemas validity error : Element \'Article\': No matching global declaration available for the validation root.' ] }
=> Meteor server restarted
=> Exited with code: 3
=> Your application is crashing. Waiting for file change.
```

**Comment by abernix** [comment]
Have you tried a simple reproduction using the direct NPM support in Meteor 1.3.2.4?

Also, please narrow your 'simple reproduction' down and post it as a git repo that can be easily checked out by developers.  As it stands, you are making several different calls that could be narrowed down more for the purposes of debugging.  As it looks to me, the `xmllint` binary is not liking the format of your file and is seg-faulting the entire node process.  This won't be properly handled by a try/catch.

**Comment by CTimmerman** [comment]
I've tried several methods, but this is the only working one that doesn't work as it should. I've removed the other attempts to provide a minimal testcase; have you executed my testcase without Meteor exiting? It appears you have, so please remove the "can't-reproduce" tag.

> seg-faulting the entire node process. This won't be properly handled by a try/catch.

So Meteor isn't supposed to handle that? Then this issue is a "wont-fix" and the problem is https://github.com/kripken/xml.js/issues/11

As mentioned in the first issue i filed (https://github.com/meteor/meteor/issues/6429) i have a workaround. (https://github.com/kripken/xml.js/issues/11#issuecomment-191158626)

**Comment by abernix** [comment]
Apologies, I just thought that was a gist link.  I still think this could have been much simplified but will see if I can get this running on my Windows box.

**Comment by abernix** [comment]
I'm able to reproduce this on Windows and OS X, so it's not isolated to Windows.  I'll leave this here as `confirmed` to see if anyone can help out, however seeing as [others are reporting the same problem running it under Node.js alone](https://github.com/kripken/xml.js/issues/11#issuecomment-217655391), I'm betting that this is a problem with `xmllint`.

**Comment by CTimmerman** [comment]
"I'm not able to reproduce this issue with node itself using any node@4.4.2/5.x/6.x" - https://github.com/kripken/xml.js/issues/11#issuecomment-218363252

**Comment by abernix** [comment]
@CTimmerman I posted a stack trace of the xmllint error that's occurring when run in the Meteor environment (taken from `node-inspector`) in that issue to see if they can help out in any way.

**Comment by hwillson** [comment]
Hi all - I've taken a look at this, and have also been able to reproduce the same problem with the provided repro. I've tested a version of the repro that uses Meteor 1.4.3.2 and Meteor's out of the box npm support, but the same problem is still occurring. I've looked at the [`xmllint`](https://github.com/kripken/xml.js) codebase briefly - there are definitely a few interesting things happening there that might be causing this issue (like the result of the [compile/embed approach](https://github.com/kripken/xml.js/blob/d710101c4b051f007af43080a835457f4cf8fc5e/script/compile) being used to get the core `xmllint.raw.js` code in place). 

The above being said, I'm not sure that digging into this further will be worth the effort. There are other npm based XML linting packages available (that also leverage `libxml`) that achieve the same functionality as `xmllint`, and don't crash Meteor. Take a look at [`libxml-xsd`](https://www.npmjs.com/package/libxml-xsd) for example. I've created a quick repro [here](https://github.com/hwillson/meteor-issue-7012) that uses `libxml-xsd` and the original repro's XML/XSD files, to show that it works properly and doesn't kill the server. Running that repro will give the following (expected) validation error:

```
I20170320-20:41:35.784(-4)? [ { [Error: Element 'Article': No matching global declaration available for the validation root.
I20170320-20:41:35.785(-4)?   ] domain: 17, code: 1845, level: 2, column: 0, line: 2 } ]
```

Given the fact that it's difficult to pinpoint the exact problem with Meteor + `xmllint` (ie. time consuming), and that alternatives exist that achieve the same functionality (using some of the same core libraries behind the scenes), I think it's safe to close this issue. There hasn't been much activity on this issue in a while, and we can always re-open this issue if anyone disagrees (but the `libxml-xsd` package really does seem to be a better solution regardless). Thanks!
